### PR TITLE
Stateless tests: Improve tests speed

### PR DESCRIPTION
**Pull request by fm4v**
**Results**
- The total number of jobs for stateless tests per CI run is reduced, from 49 to 28.
- The Stateless tests (release) job now runs in 40 minutes instead of 60 minutes.
- The number of parallel testing jobs in a single test run is decreased, so I expect tests to be more stable with fewer timeouts.

### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Stateless tests: Improve tests speed and decrease number of parallel jobs

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [x]  Allow: All Required Checks
- [x]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: Normal Builds
- [ ]  Allow: Special Builds
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [x]  Exclude: All with ASAN
- [x]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [x]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit b653037afc3ef22f648c58d1e9a40bf49addf4f9 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65186/b653037afc3ef22f648c58d1e9a40bf49addf4f9/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: tests/queries/0_stateless/00623_truncate_all_tables.sql
Comment (Algunenano) [inline tests/queries/0_stateless/00623_truncate_all_tables.sql:1]: I guess this test was setup as no-parallel because it creates a database, so if it's run in parallel with itself it will fail.

We could change this test to a `.sh` and use `$CURRENT_DATABASE` in the `TRUNCATE ALL TABLES` call, removing the need of creating and dropping a database manually, which is good for performance and consistency (and we can remove all the `truncate_test.`)

File: tests/queries/0_stateless/01036_no_superfluous_dict_reload_on_create_database.sql
Comment (Algunenano) [inline tests/queries/0_stateless/01036_no_superfluous_dict_reload_on_create_database.sql:1]: Same here. Creating a named database isn't parallel safe.

File: tests/queries/0_stateless/01129_dict_get_join_lose_constness.sql
Comment (Algunenano) [inline tests/queries/0_stateless/01129_dict_get_join_lose_constness.sql:5]: We probably should not be creating thrash under `system` to begin with (unless we are testing creating trash under system explicitly).

**Review by Algunenano — CHANGES_REQUESTED** [review]
I haven't reviewed everything because I see it's in progress, but you are removing many `no-parallel` tags that are there because the test creates and drops a database and that's not safe to run in parallel with itself.

Most of them are probably added because otherwise the flaky check will detect this problem, and also most of them do not need to create a database and should be using the random one created for each test. This can be done in 3 ways:
* Do not do anything.
* Use currentDatabase() when you need to reference the string
* Use bash and $CURRENT_DATABASE when you need it explicitly in a query. The reference file will replace it by `default` when checking the results.

**Comment by fm4v** [comment]
@Algunenano Yes, I'm trying to remove unnecessary `no-parallel` tags, and it is safe to create a database in the test if the same database name is not used by other tests. However, it appears that the flaky check fails, but the regular test run is correct.

**Comment by Algunenano** [comment]
> However, it appears that the flaky check fails, but the regular test run is correct.

Yes, the flaky test will fail which is bad, because it will run every time we modify the test, making the check worthless. Same if you run the test locally many times.

Maybe it could be worth to slowly remove the creation of databases on those tests and making them truly parallel

File: tests/queries/0_stateless/00623_truncate_all_tables.sql
Comment (tavplubix) [reply tests/queries/0_stateless/00623_truncate_all_tables.sql:1]: It's not necessary to rewrite in bash: https://github.com/ClickHouse/ClickHouse/blob/3fc65a68c7cd6aabea5250784ee068014ffaa801/tests/queries/0_stateless/00080_show_tables_and_system_tables.sql#L2

**Comment by fm4v** [comment]
To improve test runs, we need to identify what takes the most time in a single test job:
- 5 min setup
- 15 min for all parallel tests
- 40 min for sequential tests

To speed up test runs, we need to make the sequential tests much faster. Tests are tagged with `no-parallel` for the following reasons:
- Usage of shared resources: databases, dictionaries from XML, users, system databases and tables, global SYSTEM commands, etc.
- High load due to multi-threaded object creation (replicas, databases, tables)
- Inability to pass the flaky check, which runs the same test 100 times in parallel, leading to the use of the no-parallel tag to pass it.

This PR implements several changes to optimize run speed, prioritized by impact on performance:
- Run sequential tests on a dedicated CH process in parallel with all other tests.
- Convert long-running tests tagged with no-parallel to parallel tests.
- Decrease the randomization of max_threads and max_insert_threads settings.
Other changes might reduce run time, but the exact impact is difficult to quantify.

**Results**
When test run time is decreased, we can reduce the number of test batches and keep the total run time under 1 hour.
- The total number of batches for stateless tests per CI run is reduced, from X to Y (TODO).
- The number of parallel testing jobs in a single test run is decreased, so I expect tests to be more stable with fewer timeouts.
- The Stateless tests (release) job now runs in 40 minutes instead of 60 minutes.
- ASAN batches are reduced from 4 to 2 with the same test run length.
- Waiting for a full CI run...

File: docker/test/stateless/run.sh
Comment (Algunenano) [inline docker/test/stateless/run.sh:112]: 2 things about this:
* It'd be nicer to be explicit and call them `/var/lib/clickhouse-sequential` or something similar, and `clickhouse-server-sequential.log` / `clickhouse-server-parallel.log`.
* In the report, we need to know which logs to check for a test, so it'd be nice to have a column that says `Parallel / Sequential` next to the Test time.

WDYT?

File: tests/clickhouse-test
Comment (Algunenano) [inline tests/clickhouse-test:753]: This feels funny, because it's detecting a situation where we are scheduling more threads than we need even if they are not necessary. Still it seems ok.

File: tests/clickhouse-test
Comment (Algunenano) [inline tests/clickhouse-test:1618]: Would it be possible to do this always? I for one have multiple servers and usually run against PORT 49000 so it'd be useful even when not running in parallel:
```
$ cat ~/.bashrc | grep "export CLICKHOUSE"
export CLICKHOUSE_PORT_HTTP=48123
export CLICKHOUSE_PORT_TCP=49000
export CLICKHOUSE_PORT_KEEPER=12181
```

File: tests/clickhouse-test
Comment (Algunenano) [inline tests/clickhouse-test:1982]: Not sure how this works, but this test should be executed in all batches, at the end and serially. Now I don't see it anywhere (in the reports)

File: tests/queries/0_stateless/01113_local_dictionary_type_conversion.sql
Comment (Algunenano) [inline tests/queries/0_stateless/01113_local_dictionary_type_conversion.sql:1]: Seems like this database is not necessary either. Just use `DB currentDatabase()` in the dictionary declaration, right?

File: tests/queries/0_stateless/01125_dict_ddl_cannot_add_column.sql
Comment (Algunenano) [inline tests/queries/0_stateless/01125_dict_ddl_cannot_add_column.sql:1]: Same, it should be possible to use the test database

File: tests/queries/0_stateless/01259_dictionary_custom_settings_ddl.sql
Comment (Algunenano) [inline tests/queries/0_stateless/01259_dictionary_custom_settings_ddl.sql:3]: Same -> currentDatabase()

File: tests/queries/0_stateless/01658_read_file_to_stringcolumn.sh
Comment (Algunenano) [inline tests/queries/0_stateless/01658_read_file_to_stringcolumn.sh:12]: Is this really necessary? I'd expect this to exist. Maybe it was meant to be a subdirectory?

**Review by Algunenano — COMMENTED** [review]
Incredible work.

I've left some comments. Please make sure to verify the sync PR too, since there are some conflicts and the CI is not running there because of it.

File: tests/clickhouse-test
Comment (fm4v) [reply tests/clickhouse-test]: I've added some randomization of these settings

File: tests/clickhouse-test
Comment (fm4v) [reply tests/clickhouse-test]: Nice idea, changed

File: tests/clickhouse-test
Comment (fm4v) [reply tests/clickhouse-test:1983]: This test is executed on single batch, so I decided to remove it from the CI multi-batch job. It still executes in the single-batch `Release` job.
The reason for this is that the test checks the threshold number of logs, and in batches, we execute far fewer tests, so fewer logs are generated. This causes the test to fail falsely due to the reduced log volume.

File: docker/test/stateless/run.sh
Comment (fm4v) [reply docker/test/stateless/run.sh:112]: Let's leave it as is and rename it later if it becomes a problem for people

**Review by Algunenano — APPROVED** [review]

File: tests/clickhouse-test
Comment (azat) [reply tests/clickhouse-test:1983]: This looks very icky in the code, maybe one more tag should be introduced to avoid this, like `no-batch` or similar

Comment (fm4v) [reply tests/clickhouse-test]: @azat yes it's better, done

**Comment by maxknv** [comment]
We have OOM quite often. Mostly with `Stateless tests (msan, distributed cache, s3 storage) [1/3]` and `Stateless tests (asan, distributed cache, s3 storage) [1/3]`. 
OOM happens with the first job batch in more than 95% of cases. It looks very suspicious. Is the first batch used for non-parallel tests now?

File: docker/test/stateless/run.sh
Comment (azat) [inline docker/test/stateless/run.sh:147]: You forgot to call `setup_logs_replication` for this instance, and while you are here, can you fix this for `USE_DATABASE_REPLICATED` as well

Comment (fm4v) [reply docker/test/stateless/run.sh:147]: @maxknv Most likely the test in this group is running out of memory, probably because this test is now running in parallel with other tests. I'll fix it.
@azat I'll will fix `setup_logs_replication` but fixing `USE_DATABASE_REPLICATED` does not make sense, since the number of these builds and the time to run the tests does not need to be optimized

Comment (azat) [reply docker/test/stateless/run.sh:147]: I meant add `setup_logs_replication` there as well

File: tests/clickhouse-test
Comment (azat) [inline tests/clickhouse-test:1455]: Why you prefer to use `\n` it became too verbose, let's get compact behavior back - https://github.com/ClickHouse/ClickHouse/pull/75530

**Pull request merged**
